bridge/slack/helpers.go

```diff
@@ -38,27 +38,54 @@ func (b *Bslack) getChannel(channel string) (*slack.Channel, error) {
 }
 
 func (b *Bslack) getChannelByName(name string) (*slack.Channel, error) {
-	if b.channels == nil {
-		return nil, fmt.Errorf("%s: channel %s not found (no channels found)", b.Account, name)
-	}
-	for _, channel := range b.channels {
-		if channel.Name == name {
-			return &channel, nil
-		}
+	if channel, ok := b.channelsByName[name]; ok {
+		return channel, nil
 	}
 	return nil, fmt.Errorf("%s: channel %s not found", b.Account, name)
 }
 
 func (b *Bslack) getChannelByID(ID string) (*slack.Channel, error) {
-	if b.channels == nil {
-		return nil, fmt.Errorf("%s: channel %s not found (no channels found)", b.Account, ID)
+	if channel, ok := b.channelsByID[ID]; ok {
+		return channel, nil
+	}
+	return nil, fmt.Errorf("%s: channel %s not found", b.Account, ID)
+}
+
+func (b *Bslack) populateUsers() {
+	users, err := b.sc.GetUsers()
+	if err != nil {
+		b.Log.Errorf("Could not reload users: %#v", err)
+		return
+	}
+	b.users = map[string]*slack.User{}
+	for _, user := range users {
+		b.users[user.ID] = &user
 	}
-	for _, channel := range b.channels {
-		if channel.ID == ID {
-			return &channel, nil
+}
+
+func (b *Bslack) populateChannels() {
+	b.channelsByID = map[string]*slack.Channel{}
+	b.channelsByName = map[string]*slack.Channel{}
+
```

Comment:
Maybe put a `Lock()` here as it's called from multiple routines? or use `sync.Map`?
Same in `getChannelByName` and  `getChannelByID` (they're getting called a lot by the `Send()` function)

Also maybe better to not empty them (`b.channelsByID` and  `b.channelsByName`) here, but create new temporary ones and swap them at the end after they're filled in to keep the locking to a minimum?

Reply:
Good call indeed. Have kept the locking to a minimum and added RW locks to keep parallel read calls to the maps possible.